### PR TITLE
RNN policy implemented with PyTorch

### PR DESCRIPTION
**Pull request by ml4046**
**Proposed changes**:
- Added a RNN policy implemented with PyTorch using a GRU cell

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [ ] added some tests for the functionality
- [x] updated the documentation
- [ ] updated the changelog

### CODE REVIEW

**Comment by harrywang** [comment]
@ml4046 can you pull down the changes from rasa_core? we are 5 commits behind.

**Comment by harrywang** [comment]
@onkarsalvi7 could you please do a quick review of this?

**Comment by onkarsalvi7** [comment]
@harrywang yes. sure.

File: rasa_core/policies/torch_policy.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+            self.current_epoch += 1
+
+    # TODO: test this function
+    def _continue_training(self, training_trackers, domain, **kwargs):
+        logger.info("Continuing training for model")
+        batch_size = kwargs.get("batch_size", 16)
+        epochs = kwargs.get("epochs", 10)
+
+        # TODO: use same optimizer from train()
+        optimizer = nn.optim.Adam(self.model.parameters(), lr=1e-3)
```
Comment (onkarsalvi7) [inline rasa_core/policies/torch_policy.py]: fix this typo. use `torch.optim.Adam`

File: rasa_core/policies/torch_policy.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        for ep in range(epochs):
+            training_data = self._training_data_for_continue_training(
+                batch_size, training_trackers, domain)
+            X, y = _convert_to_tensors(training_data)
```
Comment (onkarsalvi7) [inline rasa_core/policies/torch_policy.py]: check the use of `_convert_to_tensors()` though it is not being used during the training process, I tried testing it and the code breaks at this point.

File: rasa_core/policies/torch_policy.py (hunk 1 above)
Comment (ml4046) [reply rasa_core/policies/torch_policy.py]: thanks for catching this

File: rasa_core/policies/torch_policy.py (hunk 2 above)
Comment (ml4046) [reply rasa_core/policies/torch_policy.py]: This is from `_continue_training()` which is optional to implement for `Policy`. This feature will be something to consider later on.

Comment (ml4046) [reply rasa_core/policies/torch_policy.py]: resolved in latest commit

**Comment by ml4046** [comment]
@ksurya this is ready to go

File: rasa_core/policies/torch_policy.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+                                                 output_size
+                                                 )
+            self.input_size = input_size
+            self.output_size = output_size
+
+        val_split = kwargs.get("validation_split", 0.1)
+        train_size = int(shuffled_X.shape[0] * val_split)
+
+        # convert to float since model requires float
+        train_X = torch.tensor(shuffled_X[:train_size]).type(torch.float)
```
Comment (ksurya) [inline rasa_core/policies/torch_policy.py]: `if shuffled_X.shape[0] is 100, val_split = 0.1, train_size will be 10. Therefore train_X will have indexes [0... 10]. I am assuming this should be [0.... 90]`

Comment (ml4046) [reply rasa_core/policies/torch_policy.py]: correct, just fixed it, thanks for the catch

File: rasa_core/policies/torch_policy.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+                          "Nothing to persist.")
+
+    @classmethod
+    def load(cls, path):
+        # type: (Text) -> TorchPolicy
+        if os.path.exists(path):
+            featurizer = TrackerFeaturizer.load(path)
+            meta_path = os.path.join(path, "torch_policy.json")
+            if os.path.isfile(meta_path):
+                with io.open(meta_path) as f:
```
Comment (ksurya) [inline rasa_core/policies/torch_policy.py]: how is `io.open()` different from `open()`?

**Comment by ksurya** [comment]
I've made some changes, will update on this branch

File: rasa_core/policies/torch_policy.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+                    meta = json.loads(f.read())
+                model_file = os.path.join(path, meta["model"])
+                model = torch.load(model_file)
+                return cls(featurizer=featurizer,
```
Comment (ksurya) [inline rasa_core/policies/torch_policy.py]: Both `TorchPolicy.__init__` and `TorchPolicy.train` functions are flexible to accept custom hyper-parameters. But which hyper-parameters should be passed at `__init__`?  At the moment, we are not persisting the arguments to these functions except `epoch`.  Let's say if we have `num-of-layers` as hyper-parameter passed through either of these functions, and we don't persist it -- there is a possibility that we use incorrect argument at inference time! Probably I am thinking wrong.. hmmm.

Comment (ml4046) [reply rasa_core/policies/torch_policy.py]: Yes this is definitely a feature that can be added to allow user to adjust these params, then `persist` will have to save params such as `num_layers`, `hidden_size` etc. I basically did a copy of how rasa handled these cases where `KerasPolicy` will always have size 32 and 1 layer so these params don't have to be persisted.

Comment (ksurya) [reply rasa_core/policies/torch_policy.py]: 1. The KerasPolicy initializes hyper-parameters using `defaults` class variable, not as arguments through `__init__`. So, `persist` function will not have a possibility of loading incorrect hyper-parameters.

File: rasa_core/policies/torch_policy.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+                    print_val_loss += loss.item()
+
+            logger.info("Epoch {} | Train loss/acc: {} / {} | "
+                        "Val loss/acc {} / {}"
+                        .format(ep + 1,
+                                round(print_loss, 3),
+                                round(train_acc / train_it, 3),
+                                round(print_val_loss, 3),
+                                round(val_acc / val_it, 3)))
+            self.current_epoch += 1
```
Comment (ksurya) [inline rasa_core/policies/torch_policy.py]: `train()` can pass any number of epochs, but `self.current_epoch += 1` would only increment current epochs by 1

Comment (ml4046) [reply rasa_core/policies/torch_policy.py]: Correct but this is to keep track of how many epochs we have trained on so that when we call `continue_training()` we know the number of epochs we have already trained on

Comment (ksurya) [reply rasa_core/policies/torch_policy.py]: The number of epochs we trained on is based on `epochs = kwargs.get("epochs", 30)`. Should that be updated there?

File: rasa_core/policies/torch_policy.py (hunk 5 above)
Comment (ml4046) [reply rasa_core/policies/torch_policy.py]: I see, I can edit it so that it isn't a parameter in `__init__`, is that good?

Comment (ksurya) [reply rasa_core/policies/torch_policy.py]: Hopefully it won't raise flags from Rasa, because its confirming with KerasPolicy's approach.

File: rasa_core/policies/torch_policy.py (hunk 6 above)
Comment (ml4046) [reply rasa_core/policies/torch_policy.py]: Yea we can just pass `self.current_epochs = epochs` but shouldn't `train()` always starts at 0 otherwise we call `continue_training()`?

Comment (ksurya) [reply rasa_core/policies/torch_policy.py]: `continue_training` is executed only in interactive mode i.e., with `rasa_core.train --interactive`. It is possible to continue training in a non-interactive mode by loading a model from file with `rasa_core.train --core <persisted-model>`.

Comment (ksurya) [reply rasa_core/policies/torch_policy.py]: I think `continue_training` function name is misleading.

**Comment by ksurya** [comment]
@ml4046 I have finished reviewing the code, post resolving the current items, please review the code to see if there are any changes required. We'll have the PR to RasaHQ

File: rasa_core/policies/torch_policy.py (hunk 6 above)
Comment (ml4046) [reply rasa_core/policies/torch_policy.py]: I see, that makes sense. But I think `self.current_epoch+=1` does the job of tracking number of epochs but just not as elegant as `self.current_epoch += epochs`. Is this what you were getting at?

Comment (ksurya) [reply rasa_core/policies/torch_policy.py]: Yes

**Comment by ml4046** [comment]
@ksurya This is ready to go.
